design/automatic-network-configuration.md

```diff
@@ -0,0 +1,738 @@
+    2. Prepares for deletion.
+    3. The state handler will remove finalizers.
+
+![](./images/automatic-network-configuration-state.png)
+
+#### SmartNIC
+
+##### SmartNICPort CRD
+
+```yaml
+apiVersion: v1alpha1
+kind: SmartNicPort
+metaData:
+  name: port2
+  ownerRef:
+    name: smart1
+  finalizers:
+  - defaulf-bm0
+spec:
+  configurationRef:
+    name: sn1
+    namespace: default
+  # Represents the next port information of this port link
+  nextPortRef:
+    name: port3
+    kind: SwitchPort
+    namespace: default
+status:
+  # Indicates the actual configuration status of the port
+  state: Configured
+  # Indicates the configuration information currently applied to the port
+  configurationRef:
+    name: sn1
+    namespace: default
+```
+
+##### SmartNicConfiguration CRD
+
+```yaml
+apiVersion: v1alpha1
+kind: SmartNicConfiguration
+metaData:
+  name: sc1
+  namespace: default
+  finalizers:
+    - default-port0
+spec:
+  ...
+  nextConfigurationRef:
+  name: sp1
+  kind: SwitchPortConfiguration
+  namespace: default
+```
+
+### Multi-tenancy models
+
+- No multi-tenancy : one set of controllers deployed, all objects are
+  reconciled by the same controllers.
+- "Limited" multi-tenancy : one set of CRs (the most up-to-date), one set of
+  webhooks and a single network controller with clusterRole and in its own
+  namespace. All other controllers (CAPI + CAPM3) per namespace. Permission
+  separation between namespaces.
+- Full multi-tenancy : one set of CRs (the most up-to-date), one set of
+  webhooks. All other controllers per namespace. No permission separation
+  within the same namespace.
+
+### Permissions
+
+The permissions for a role include the permissions given to the controllers
+deployed by that role.
+
+#### Scenario: One cluster operator
+
+Description: Only one cluster operator has full management access to all resources
+
+Multi-tenancy models:
+
+- None
+
+Roles:
+
+- Operator: Full control
+  Scope: Cluster
+
+Operations Allowed:
+||Operation / resource|Goal|
+|:-|:-|:-|
+|Operator|All/All|Configure everything|
+
+Operations Not Allowed:
+||Operation|Goal|
+|:-|:-|:-|
+|Operator|None||
+
+#### Scenario: Many cluster operators
+
+Description: Each operator has full management access to all resources in his namespace.
+
+Multi-tenancy models:
+
+- Full
+
+Roles:
+
+- Operator: Full control
+  Scope: Cluster
+
+Operations Allowed:
+||Operation / resource|Goal|
+|:-|:-|:-|
+|Operator|All/All|Configure everything|
+
+Operations Not Allowed:
+||Operation|Goal|
+|:-|:-|:-|
+|Operator|None||
+
+#### Scenario: Infra team and consumers
+
+Description: There is an owner of the infra and some consumers. The infra
+team is the owner of the management cluster and manages the physical resources.
+A consumer is the manager of the target cluster, responsible for its lifecycle.
+The consumer gets resources from the infra team (tenant). No consumer action
+should impact another consumer. Metal3 setup on a shared management cluster,
```

Comment:
I agree that no consumer action should impact another consumer, but the rules below don't enforce that because they assume that the only resource that is exclusive to a consumer is the Port.
In reality, the common multitenancy use case would be to assign different VLANs to different consumers. Allowing each consumer to configure their own ports in any way they want would permit them to connect their hosts to any VLAN.

Reply:
This is indeed a problem that we did not consider, and we intend to solve it in the following ways:
Each user can specify any vlan. If this vlan is used by other users in the switch, it will randomly find an unused vlan and put the port in this new vlan. But the user does not know, only the operator knows and records this information. This allows different users to specify the same vlan, and the actual configuration is not in the same vlan. WDYT?

Reply:
I would very strongly advise against that.

My intention here was to point out one of the ways in which the current model fails to account for multitenancy _of the network_ as discussed above. (I suspect this is also what @maelk was getting at when he [originally suggested](https://github.com/metal3-io/metal3-docs/pull/132#discussion_r485369047) documenting the approach to RBAC.) We should look at the problem in that context.

Reply:
I could imagine that being useful, although the config needs to be in a CRD, not in a file.

Reply:
@zaneb does the current version address your concerns? I assume the configuration-port split does it?

Reply:
@zaneb The main remaining problem is this. Do you think our current design can solve this problem?
Can you and @dtantsur review again? If there are no other questions, can you give an approval? This is very important to us, thank you!